cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowImpl.qll

```diff
@@ -1998,65 +1982,58 @@ private predicate flow0(NodeExt node, boolean toReturn, AccessPath ap, Configura
     flow(mid, toReturn, ap, config)
   )
   or
-  exists(NodeExt mid, AccessPathNil ap0 |
-    argumentFlowsThrough3(node, mid, _, _, TSummaryTaint(), config) and
-    flow(mid, toReturn, ap0, config) and
+  exists(NodeExt mid, AccessPathNil nil |
+    argumentFlowsThrough(node, mid, _, config) and
+    flow(mid, toReturn, nil, config) and
     ap instanceof AccessPathNil and
     flowFwd(node, _, _, ap, config)
   )
   or
-  exists(NodeExt mid, AccessPath ap0 |
-    storeFwd(node, _, mid, ap, ap0, config) and
-    flow(mid, toReturn, ap0, config)
-  )
-  or
-  exists(Content f, AccessPath ap0 |
-    flowTaintStore(node, f, toReturn, ap0, config) and
-    pop(f, ap0) instanceof AccessPathNil and
-    ap instanceof AccessPathNil and
-    flowFwd(node, _, _, ap, config)
+  exists(Content f |
+    flowStore(f, node, toReturn, ap, config) and
+    flowConsCand(f, ap, config)
   )
   or
   exists(NodeExt mid, AccessPath ap0 |
-    readFwd(node, mid, ap, ap0, config) and
+    readFwd(node, _, mid, ap, ap0, config) and
```

Comment:
I think we also need to join with `flowConsCandFwd(f, _, ap0, config)` here.  Alternatively, the conjunct should be pushed into `readFwd` if it doesn't yield a bad join-order.

Reply:
I _think_ this may yield a slight reduction in the number of tuples in `flow`.

Reply:
Good idea 👍